### PR TITLE
Feature/handle request errors

### PR DESCRIPTION
**Pull request by rrrrs09**

### CODE REVIEW

**Review by YunusovSamat — APPROVED** [review]

**Review by Nimond — APPROVED** [review]

File: botx/exceptions.py [hunk 1]
```diff
@@ -83,3 +83,32 @@ class TokenError(BotXException):
...
+
+class BotXConnectError(BotXException):
+    """Raised if unable to connect to service."""
+
+    message_template = " ".join(
+        [
+            "unable to connect to service {method} {url}.",
+            "Make sure you specified the correct host in bot credentials.",
+        ],
+    )
```
Comment (vonabarak) [inline botx/exceptions.py:108]:
```suggestion
    message_template = (
        "unable to connect to service {method} {url}."
        "Make sure you specified the correct host in bot credentials."
    )
```

File: tests/test_clients/test_clients/test_async_client/test_execute.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+import pytest
+from httpx import ConnectError, Request, Response
+
+from botx.clients.methods.v2.bots.token import Token
+from botx.exceptions import BotXConnectError, BotXJSONDecodeError
+
+try:
+    from unittest.mock import AsyncMock
+except ImportError:
+    from unittest.mock import MagicMock
```
Comment (vonabarak) [inline tests/test_clients/test_clients/test_async_client/test_execute.py:12]: Mb add some comment that this code here is for 3.7 compatibility

File: botx/clients/clients/async_client.py [hunk 3]
```diff
@@ -90,16 +96,28 @@ async def execute(self, request: HTTPRequest) -> HTTPResponse:
...
-        )
+        try:
+            response = await self.http_client.request(
+                request.method,
+                request.url,
+                headers=request.headers,
+                params=request.query_params,
+                json=request.json_body,
+            )
+        except httpx.HTTPError:
```
Comment (lxmnk) [inline botx/clients/clients/async_client.py:112]:
```python
except httpx.HTTPError as httpx_exc:
    raise BotXConnectError(url=request.url, method=request.method) from httpx_exc
```

**Review by vonabarak — APPROVED** [review]

File: botx/clients/clients/async_client.py [hunk 4]
```diff
@@ -90,16 +96,28 @@ async def execute(self, request: HTTPRequest) -> HTTPResponse:
...
+            raise BotXConnectError(url=request.url, method=request.method)
+
+        try:
+            json_body = response.json()
+        except JSONDecodeError:
```
Comment (lxmnk) [inline botx/clients/clients/async_client.py:117]: Same here.

File: botx/exceptions.py (hunk 1 above)
Comment (rrrrs09) [reply botx/exceptions.py:108]: It doesn't work because of linters.

File: tests/test_clients/test_methods/test_base/test_empty_error_handlers.py [hunk 5]
```diff
@@ -0,0 +1,13 @@
+from botx.clients.methods.base import BotXMethod
+
+
+class TestMethod(BotXMethod):
+    __url__ = "/path/to/example"
+    __method__ = "GET"
+    __returning__ = str
+
+
+def test_method_empty_error_handlers():
```
Comment (lxmnk) [inline tests/test_clients/test_methods/test_base/test_empty_error_handlers.py:10]: Nit: Move the class inside the function.

File: botx/clients/methods/errors/unauthorized_bot.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
+"""Definition for "invalid signature" error."""
+from typing import NoReturn
+
+from botx.clients.methods.base import APIErrorResponse, BotXMethod
+from botx.clients.types.http import HTTPResponse
+from botx.exceptions import BotXAPIError
+
+
+class InvalidSignatureError(BotXAPIError):
```
Comment (lxmnk) [inline botx/clients/methods/errors/unauthorized_bot.py:9]: `InvalidSignatureError` -> `InvalidBotCredentials`?

File: botx/clients/clients/async_client.py (hunk 3 above)
Comment (rrrrs09) [reply botx/clients/clients/async_client.py]: Fixed

File: botx/exceptions.py (hunk 1 above)
Comment (rrrrs09) [reply botx/exceptions.py]: Done

File: tests/test_clients/test_clients/test_async_client/test_execute.py (hunk 2 above)
Comment (rrrrs09) [reply tests/test_clients/test_clients/test_async_client/test_execute.py:12]: Added

File: botx/clients/clients/async_client.py (hunk 4 above)
Comment (rrrrs09) [reply botx/clients/clients/async_client.py]: Fixed

File: tests/test_clients/test_methods/test_base/test_empty_error_handlers.py (hunk 5 above)
Comment (rrrrs09) [reply tests/test_clients/test_methods/test_base/test_empty_error_handlers.py:10]: mypy raised internal error when I moved

File: botx/clients/methods/errors/unauthorized_bot.py (hunk 6 above)
Comment (rrrrs09) [reply botx/clients/methods/errors/unauthorized_bot.py]: Done

**Review by vonabarak — APPROVED** [review]

File: tests/test_clients/test_methods/test_base/test_empty_error_handlers.py (hunk 5 above)
Comment (lxmnk) [reply tests/test_clients/test_methods/test_base/test_empty_error_handlers.py:10]: Hmm, ok.

**Review by lxmnk — APPROVED** [review]

**Review by Opti213 — APPROVED** [review]

**Pull request merged**